### PR TITLE
Support for docking & maximizing through pointer/mouse movement

### PR DESCRIPTION
**Pull request by AlanGriffiths**
Drag left/right docks left/right
Drag top maximizes

Fixes: #109

### CODE REVIEW

**Comment by AlanGriffiths** [comment]
> The problem I described above must've been a bug. Unless my window was big enough to annoy Miriway? When restored, it was basically screen size, so if the logic is about where the _window_ is, that kicked it about?

The window probably overlapped the side(s).

> It should all be about the cursor position when moving, no? If sufficiently near the edge, dock them?

Any thoughts on what would be a good "near" would be? 10% of screen width/height? Without visual feedback that may also be tricky to use

**Comment by Saviq** [comment]
> Any thoughts on what would be a good "near" would be? 10% of screen width/height? Without visual feedback that may also be tricky to use

On "outer" edges it should be minimal. On display boundary something like 2-5%?

Actually, it should be both - cursor near the edge _and_ window going over the edge. That would help a lot with accidental docking? It'd be somewhat fringe to expect to dock a window that you grabbed at its very edge, such that the cursor is within the threshold, but the window isn't? That should be fine without a visual indicator I think.

**Comment by AlanGriffiths** [comment]
Adding an additional criteria for docking seems to be in conflict with your first impression: "I have to press really hard against the top edge to get it to maximize, could use a bigger threshold."

**Comment by Saviq** [comment]
> Adding an additional criteria for docking seems to be in conflict with your first impression: "I have to press really hard against the top edge to get it to maximize, could use a bigger threshold."

Not if one of the criteria gets relaxed? AIUI it was "be at/over the edge"?

That the window overlaps an edge is too easy to trigger by mistake. If it's also needed that you drop it with the cursor near the edge (but not necessarily _at_ the edge) should be a good balance?

**Comment by AlanGriffiths** [comment]
> > Adding an additional criteria for docking seems to be in conflict with your first impression: "I have to press really hard against the top edge to get it to maximize, could use a bigger threshold."
> 
> Not if one of the criteria gets relaxed? AIUI it was "be at/over the edge"?

Which criteria are you thinking to relax? Currently we have "window over the edge", I think you're suggesting adding "AND cursor in margin". (I don't think the additional pixel out of possibly several thousand from "at the edge" is a meaningful relaxation.)

**Comment by Saviq** [comment]
> Currently we have "window over the edge"

Hmm this wasn't what I experienced, I had to "press" against the edge as well quite… hard. Let me try again.

If that is the case, then that sounds too easy to trigger. Sometimes you just want to move the window out of the way, and if that happens to cross the edge, it will dock.

**Comment by AlanGriffiths** [comment]
> If that is the case, then that sounds too easy to trigger. Sometimes you just want to move the window out of the way, and if that happens to cross the edge, it will dock.

OK. The latest iteration only docks if the window overlaps the edge AND the cursor is within a 5% margin of the edge. Seems to work as expected.
